server/src/main/java/org/elasticsearch/index/IndexSettings.java

```diff
@@ -72,6 +73,8 @@
             (value) -> Translog.Durability.valueOf(value.toUpperCase(Locale.ROOT)), Property.Dynamic, Property.IndexScope);
     public static final Setting<Boolean> INDEX_WARMER_ENABLED_SETTING =
         Setting.boolSetting("index.warmer.enabled", true, Property.Dynamic, Property.IndexScope);
+    public static final Setting<Integer> INDEX_MAX_DOC_ID_LENGTH_SETTING =
+        Setting.intSetting("index.max_doc_id_length", 512, 1, Property.Dynamic, Property.IndexScope);
```

Comment:
Using `doc_id` is unfortunate since the lucene "doc id" is something else, suggest to just stick to `id`:

```suggestion
        Setting.intSetting("index.max_id_length", 512, 1, Property.Dynamic, Property.IndexScope);
```

(also rename various fields and variables).